### PR TITLE
Automatic build of documentation on merge

### PR DESCRIPTION
**Pull request by dalg24**

### CODE REVIEW

File: .jenkins [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+                }
+            }
+            environment {
+                GIT_CREDS = credentials('0f170146-bd5e-44c0-a6e0-4f6fac3230fa')
+            }
+            steps {
+                sh 'apt update && apt install --no-install-recommends -y build-essential cmake doxygen git'
+                sh 'git config http.sslVerify false'
+                sh '''rm -rf build_doc && mkdir -p build_doc && cd build_doc &&
+                      cmake -D Tasmanian_ENABLE_DOXYGEN=ON .. && make Tasmanian_doxygen &&
```
Comment (mkstoyanov) [inline .jenkins]: Minor point, here we should add `-D DOXYGEN_INTERNAL_DOCS=ON`
```
cmake -D DOXYGEN_INTERNAL_DOCS=ON -D Tasmanian_ENABLE_DOXYGEN=ON; ...
```

Comment (dalg24) [reply .jenkins]: Done

File: .jenkins [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            }
+            steps {
+                sh 'apt update && apt install --no-install-recommends -y build-essential cmake doxygen git'
+                sh '''rm -rf build_doc && mkdir -p build_doc && cd build_doc &&
+                      cmake -DTasmanian_ENABLE_DOXYGEN=ON -DDOXYGEN_INTERNAL_DOCS=ON .. &&
+                      make Tasmanian_doxygen &&
+                      git fetch origin gh-pages && git branch -D gh-pages &&
+                      git checkout -b gh-pages FETCH_HEAD &&
+                      rm -rf ../rtd && mv Doxygen/html ../rtd && git add --all ../rtd &&
+                      git config http.sslVerify false
```
Comment (dalg24) [inline .jenkins]: I didn't manage to get it to work without that...

File: .jenkins [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                GIT_CREDS = credentials('0f170146-bd5e-44c0-a6e0-4f6fac3230fa')
+            }
+            steps {
+                sh 'apt update && apt install --no-install-recommends -y build-essential cmake doxygen git'
+                sh '''rm -rf build_doc && mkdir -p build_doc && cd build_doc &&
+                      cmake -DTasmanian_ENABLE_DOXYGEN=ON -DDOXYGEN_INTERNAL_DOCS=ON .. &&
+                      make Tasmanian_doxygen &&
+                      git fetch origin gh-pages && git branch -D gh-pages &&
+                      git checkout -b gh-pages FETCH_HEAD &&
+                      rm -rf ../rtd && mv Doxygen/html ../rtd && git add --all ../rtd &&
```
Comment (dalg24) [inline .jenkins]: Note that this will publish under https://ornl.github.io/TASMANIAN/rtd/
We can symlink rtfd to rtd :)

File: .jenkins [hunk 4]
```diff
@@ -0,0 +1,35 @@
+pipeline {
+    agent {
+        label 'DoNotRunAnythingOnMaster'
+    }
+    stages {
+        stage("Deploy documentation") {
```
Comment (dalg24) [inline .jenkins]: I would like to migrate the other builds to this jenkinsfile.  I plan on doing that later if you agree that it is better this way.

File: .jenkins (hunk 3 above)
Comment (mkstoyanov) [reply .jenkins]: Why delete the branch and why the subfolder? I guess I'm missing something, why not
```
# assuming we start at master and we have build added to .gitignore
mkdir -p build; cd build
cmake -D DOXYGEN_INTERNAL_DOCS=ON -D Tasmanian_ENABLE_DOXYGEN=ON ..
make Tasmanian_doxygen
cd ..; git checkout gh-pages; # checkout gh-pages then delete the current page
git rm -r *.map *.md5 *.png *.html search/ *.js *.css formula.repository
cp -r build/Doxygen/html/* .
git add *
git commit -m "new page" # added the new page
git reset --soft HEAD^^; git commit -m "updated page" # squash the last two commits
git push -f                                                                       # force push
```
That way `gh-pages` has no source, only the web-page, and the `index.html` is at the root folder so the address is:  (https://ornl.github.io/)[https://ornl.github.io/]

We can add the `RTFD` on the title on the first page :)

File: .jenkins [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+                docker {
+                    image "ubuntu:18.04"
+                    label 'docker'
+                }
+            }
+            environment {
+                GIT_CREDS = credentials('0f170146-bd5e-44c0-a6e0-4f6fac3230fa')
+            }
+            steps {
+                sh 'apt update && apt install --no-install-recommends -y build-essential cmake doxygen git'
```
Comment (mkstoyanov) [inline .jenkins]: We also need `graphviz` to get the Doxygen dot module for the formulas.

Also, are we rebuilding the docker image on every pull-request? Seems slow and wasteful.

File: .jenkins (hunk 4 above)
Comment (mkstoyanov) [reply .jenkins]: Are we rebuilding the images on every pull request? Otherwise I'm OK with the move.

Comment (mkstoyanov) [reply .jenkins]: Are we rebuilding the images on every pull request? Otherwise I'm OK with the move.

File: .jenkins (hunk 3 above)
Comment (dalg24) [reply .jenkins]: I would like to have the option to use the root page (https://ornl.github.io/TASMANIAN/) for something else, possibly a home page.

File: .jenkins (hunk 4 above)
Comment (dalg24) [reply .jenkins]: Images will be cached and rebuilt if something changed.  That's how we do it on ArborX and Kokkos now.

File: .jenkins (hunk 5 above)
Comment (dalg24) [reply .jenkins]: I was too lazy to define a Dockerfile but I will add it

Comment (dalg24) [reply .jenkins]: Done.

File: .jenkins (hunk 3 above)
Comment (mkstoyanov) [reply .jenkins]: OK. I would like to use `/rolling/` for this page, in the future we can have `/stable/` for the latest stable release (or even `/stableX.Y/` although we wouldn't have more than one).

Right now both stable and the home-page are under `tasmanian.ornl.gov` which is good for as long as we can keep it.

**Pull request merged**